File: PopularShows/app/src/main/java/com/lgcode/popularshows/views/TvShowListView.java
package com.lgcode.popularshows.views;

import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.RelativeLayout;

import com.squareup.otto.Bus;

import butterknife.BindView;
import butterknife.ButterKnife;
import glb.lg.popularshows.R;

import com.lgcode.popularshows.adapters.TVShowsAdapter;

/**
 * Created by leojg on 11/5/16.
 */
public class TvShowListView extends ActivityView {

    @BindView(R.id.loading_panel) RelativeLayout loadingPanel;
    @BindView(R.id.shows_list) RecyclerView recyclerView;

    private Bus bus;
    private boolean isLoading;
    private boolean isLastPage;
    private int pageSize;

    public TvShowListView(AppCompatActivity activity, final Bus bus) {
        super(activity);

        this.bus = bus;

        ButterKnife.bind(this, activity);

        //recyclerView = (RecyclerView) activity.findViewById(R.id.shows_list);
        recyclerView.setLayoutManager(new LinearLayoutManager(this.getActivity()));
        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);

                LinearLayoutManager layoutManager = (LinearLayoutManager) recyclerView.getLayoutManager();

                int visibleItemCount = layoutManager.getChildCount();
                int totalItemCount = layoutManager.getItemCount();
                int firstVisibleItemPosition = layoutManager.findFirstVisibleItemPosition();

                if (!isLoading && !isLastPage) {
                    if ((visibleItemCount + firstVisibleItemPosition) >= totalItemCount
                            && firstVisibleItemPosition >= 0
                            && totalItemCount >= pageSize) {

                        loadingPanel.setVisibility(View.VISIBLE);
                        isLoading = true;
                        bus.post(new TVShowsNewPageRequestEvent());
                    }
                }
            }
        });

    }

    public void setAdapter(TVShowsAdapter adapter) {
        recyclerView.setAdapter(adapter);
    }

    public void updateList(Integer currentPageSize) {
        recyclerView.getAdapter().notifyDataSetChanged();

        loadingPanel.setVisibility(View.GONE);
        isLoading = false;
        pageSize = currentPageSize;
    }

    //Events
    public class TVShowsNewPageRequestEvent {
    }

}

File: KtPopularShows/app/src/main/java/example/lgcode/ktpopularshows/mvp/view/TVShowListView.kt
package example.lgcode.ktpopularshows.mvp.view

import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.View
import example.lgcode.ktpopularshows.adapter.TVShowsAdapter
import example.lgcode.ktpopularshows.domain.TVShow
import kotlinx.android.synthetic.main.activity_show_list.*
import java.lang.ref.WeakReference

class TVShowListView(
        val activity: AppCompatActivity
): ActivityView(WeakReference(activity)) {

    fun setUp() {
        activity.shows_list.layoutManager = LinearLayoutManager(activity)
    }

    fun getListLinearLayoutManager(): LinearLayoutManager {
        return activity.shows_list.layoutManager as LinearLayoutManager
    }

    fun setShowListScrollListener(scrollListener: RecyclerView.OnScrollListener) {
        activity.shows_list.addOnScrollListener(scrollListener)
    }

    fun setAdapter(adapter: TVShowsAdapter) {
        activity.shows_list.adapter = adapter
    }

    fun updateList(tvShows: List<TVShow>) {
        (activity.shows_list.adapter as TVShowsAdapter).addAll(tvShows)
        activity.shows_list.adapter.notifyDataSetChanged()
    }

    fun showLoader() {
        activity.loading_panel.visibility = View.VISIBLE
    }

    fun hideLoader() {
        activity.loading_panel.visibility = View.GONE
    }

}